README.md
nload
=====

Fork from nload by Roland Riegel.

Auto dynamic scaling to show incoming and outcoming speed, setting for -i max_scaling -o max_scaling is invaild now.



## Install

./configure

make

make install 

more details, read the README manual


ChangeLog
2013-12-02 Liuyang1 <liuyang1@mail.ustc.edu.cn>

	* fixed option window invaild read or write error
	
	* add dynamic scaling feature

2012-02-06 Roland Riegel <feedback@roland-riegel.de>

	* Newer Linux kernels use 64 bit counters, which caused traffic to increase by 4GB on every refresh (see Debian bug #651848)

2011-06-11 Roland Riegel <feedback@roland-riegel.de>

	* Fix calculation of traffic and data units (thanks to Kirby Zhou, see also Debian bug #586943)

	* Fix memory leak in BSD network device detection

	* Workaround for faulty newwin() in ncurses 5.8

	* Correctly escape minus signs in man page

	* Tag 0.7.3 release

2008-06-28 Roland Riegel <feedback@roland-riegel.de>

	* Really fix compilation with g++ 4.3

	* Tag 0.7.2 release

2008-02-27 Roland Riegel <feedback@roland-riegel.de>

	* If devices are given on the command line, display only these

	* Use --sysconfdir as given to configure

	* Tag 0.7.1 release

2008-02-02 Roland Riegel <feedback@roland-riegel.de>

	* Abort configure script if ncurses is not found

	* Rewrite the linux proc devreader

	* Implement hpux device detection

	* Implement reading and writing of config files (fixes debian bug #247030)

	* Add F5 and F6 key shortcuts for reading and writing current settings

	* Change defaults for traffic units to human readable

	* Make shown devices configurable at run time and save this setting as well

	* Remove "-t 0"

	* Change key shortcuts, especially F2 for the option window

	* Replace "average smoothness" by average window period in seconds

	* Apply patch by Paul Brook <paul@nowt.org> so that nload works correctly on 64-bit kernels

	* Fix SIGSEGV when /proc/net/dev doesn't exist

	* Make compilation work with GCC 4.3 (fixes debian bug #417449)

	* Tag 0.7.0 release

2003-12-18 Roland Riegel <feedback@roland-riegel.de>

	* Ported to HP-UX (thanks to Roshan Sequeira)

	* Added some useful key shortcuts

	* Fixed some memory leaks (also on BSD), uninitialized variables

	* Fixed compilation with gcc 3.x

	* Source code cleanup

	* Tagged 0.6.0 release

2002-08-11 Roland Riegel <feedback@roland-riegel.de>

	* New option window (via the 'o' key) to change program settings at run time

	* Display IP numbers of the devices

	* Faster keyboard reaction

	* "Real" average calculation

	* Big code reorganisation

	* New nload-user mailing list (see README)

	* Tagged 0.5.0 release

2001-12-06 Roland Riegel <feedback@roland-riegel.de>

	* Ported to Free/Net/OpenBSD and Solaris

	* Switches -u and -U for customizable data units

	* Tripled graph resolution by adding two additional drawing chars (thanks to Christian Hoenig)

	* Fixed wrong data display after network device restart (thanks to Christian Hoenig)

	* Tagged 0.4.0 release

2001-10-20 Roland Riegel <feedback@roland-riegel.de>

	* Now the 4GB workaround really works

	* Tagged 0.3.2 release

2001-10-17 Roland Riegel <feedback@roland-riegel.de>

	* Parameter -m introduced: don't show graphs

	* Added the nload(1) man page

	* Workaround for the 4GB kernel variable overflow

	* Tagged 0.3.1 release

2001-09-25 Roland Riegel <feedback@roland-riegel.de>

	* Two-dimensional graphs for current net load

	* Arrow keys to switch between devices

	* 'q' key to exit

	* Tagged 0.3.0 release

2001-08-18 Roland Riegel <feedback@roland-riegel.de>

	* Improved average calculation

	* Warning if device doesn't exist

	* More command line parsing fixes

	* Code cleanups

	* Tagged 0.2.2 release

2001-08-08 Roland Riegel <feedback@roland-riegel.de>

	* Added average calculation

	* Fixed min/max calculation

	* Fixed severe timing/calculation problems (thanks to Magnus Roth)

	* Fixed a command line parsing bug

	* Tagged 0.2.1 release

2001-08-03 Roland Riegel <feedback@roland-riegel.de>

	* Switched to ncurses

	* Support of more than one network device

	* Displays min/max bandwidth usage

	* Tagged 0.2.0 release

2001-07-28 Roland Riegel <feedback@roland-riegel.de>

	* Seg fault when specifying a network device fixed

	* Tagged 0.1.1 release

2001-07-27 Roland Riegel <feedback@roland-riegel.de>

	* First public release 0.1.0



src/graph.h
/***************************************************************************
                                  graph.h
                             -------------------
    begin                : Sat Sep 29 2001
    copyright            : (C) 2001 - 2012 by Roland Riegel
    email                : feedback@roland-riegel.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifndef GRAPH_H
#define GRAPH_H

#include <list>

class Window;

const unsigned long long kilo = 1024;
const unsigned long long mega = 1024 * 1024;
const unsigned long long giga = 1024 * 1024 * 1024;

static unsigned long long ThresArray[]=
{0, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
	1*kilo, 2*kilo, 4*kilo, 8*kilo, 16*kilo, 32*kilo, 64*kilo, 128*kilo, 256*kilo, 512*kilo,
	1*mega, 2*mega, 4*mega, 8*mega, 16*mega, 32*mega, 64*mega, 128*mega, 256*mega, 512*mega,
	1*giga, 2*giga, 4*giga, 8*giga, 16*giga, 32*giga, 64*giga, 128*giga, 256*giga, 512*giga,
	};

static char ThresStrBit[][10]=
{"NULL","8Bit/s","16Bit/s","32Bit/s","64Bit/s","128Bit/s","256Bit/s","512Bit/s",
	"1KBit/s","2KBit/s","4KBit/s","8KBit/s","16KBit/s","32KBit/s","64KBit/s","128KBit/s","256KBit/s","512KBit/s",
	"1MBit/s","2MBit/s","4MBit/s","8MBit/s","16MBit/s","32MBit/s","64KBit/s","128KBit/s","256MBit/s","512MBit/s",
	"1GBit/s","2GBit/s","4GBit/s","8GBit/s","16GBit/s","32GBit/s","64KBit/s","128KBit/s","256GBit/s","512GBit/s",
	"1TBit/s","2TBit/s","4TBit/s",
	};
	
inline unsigned long long getThresh(unsigned long long val)
{
	unsigned int len = sizeof(ThresArray) / sizeof(unsigned long long);
	for (unsigned int i=0; i < len; i++){
		if (val <= ThresArray[i])
			return ThresArray[i];
	}
	return 0;
}
inline char* getString(unsigned long long val)
{
	unsigned int len = sizeof(ThresArray);
	for (unsigned int i=0; i<len; i++){
		if (val <= ThresArray[i]){
			return ThresStrBit[i];
		}
	}
	return ThresStrBit[0];
}

class Graph
{
    public:
        Graph();
        ~Graph();
        
        void setNumOfBars(unsigned int numOfBars);
        void setHeightOfBars(unsigned int heightOfBars);
        void setMaxDeflection(unsigned long long maxDeflection);
		unsigned long long getMaxDeflection()
		{
			return m_maxDeflection;
		}
        
        void update(unsigned long long value);
        void print(Window& window, int x, int y);
        void resetTrafficData();
		void updateMaxDeflection();
		void setDynamicMax() {
			isDynamicMax = true;
		}

    private:
		bool isDynamicMax;
        unsigned int m_heightOfBars;
        unsigned long long m_maxDeflection;
        std::list<unsigned long long> m_values;
};

#endif


src/graph.cpp
/***************************************************************************
                                 graph.cpp
                             -------------------
    begin                : Sat Sep 29 2001
    copyright            : (C) 2001 - 2012 by Roland Riegel
    email                : feedback@roland-riegel.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "graph.h"
#include "setting.h"
#include "settingstore.h"
#include "window.h"
#include <algorithm>

using namespace std;

Graph::Graph()
    : m_heightOfBars(5), m_maxDeflection(1024*1024/8)
{
	setDynamicMax();
}

Graph::~Graph()
{
}

// sets the number of the graph's vertical #-bars
void Graph::setNumOfBars(unsigned int numOfBars)
{
    // vertically resize the graph's value list
    m_values.resize(numOfBars);
}

// sets the height of the graph's vertical #-bars
void Graph::setHeightOfBars(unsigned int heightOfBars)
{
    m_heightOfBars = heightOfBars;
}

void Graph::setMaxDeflection(unsigned long long maxDeflection)
{
    m_maxDeflection = maxDeflection;
}

// new traffic measurement has been made => update the graph's value list
void Graph::update(unsigned long long value)
{
    // [new_value] = Bytes/s
    
    // put new value to the beginning of the list, it becomes the first #-bar
    m_values.push_front(value);

    // delete the last #-bar of the list, but keep at least one
    if(m_values.size() > 1)
        m_values.pop_back();
}
void Graph::updateMaxDeflection()
{
	if (isDynamicMax) {
		unsigned long long maxTraffic = *(max_element(m_values.begin(), m_values.end()));
		unsigned long long maxThresh = getThresh(maxTraffic);
		if (maxThresh != 0)
			setMaxDeflection(maxThresh);
	}
}
// print the graph with the upper left corner at the coordinates (x, y)
void Graph::print(Window& window, int x, int y)
{
	updateMaxDeflection();
    window.setXY(x, y);
	
    // cycle through through the lines
    for(unsigned int l = 0; l < m_heightOfBars; l++)
    {
        // for each line cycle through the rows
        for(list<unsigned long long>::reverse_iterator r = m_values.rbegin(); r != m_values.rend() ; r++)
        {
            unsigned long long trafficPerLine = m_maxDeflection / m_heightOfBars;
            unsigned long long lowerLimit = m_maxDeflection * (m_heightOfBars - l - 1) / m_heightOfBars;

            if(*r < lowerLimit)
            {
                window.print(' ');
            }
            else 
            {
                unsigned long long restOfTraffic = *r - lowerLimit;

                if(restOfTraffic >= trafficPerLine)
                    window.print('#');
                else if(restOfTraffic >= trafficPerLine * 7 / 10)
                    window.print('|');
                else if(restOfTraffic >= trafficPerLine * 3 / 10)
                    window.print('.');
                else
                    window.print(' ');
            }
        }
        window.print('\n');
        window.setX(x);
    }
    
}

// reset all traffic values in the graph to zero
void Graph::resetTrafficData()
{
    int size = m_values.size();
    m_values.clear();
    m_values.resize(size);
}



src/main.cpp
/***************************************************************************
                                  main.cpp
                             -------------------
    begin                : Wed Jul 25 2001
    copyright            : (C) 2001 - 2012 by Roland Riegel 
    email                : feedback@roland-riegel.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
/*
 * nload
 * real time monitor for network traffic
 * Copyright (C) 2001 - 2012 Roland Riegel <feedback@roland-riegel.de>
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
    #include <config.h>
#endif

#include "device.h"
#include "devreader.h"
#include "devreaderfactory.h"
#include "graph.h"
#include "main.h"
#include "screen.h"
#include "setting.h"
#include "settingfilter.h"
#include "settingstore.h"

#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include <ctype.h>
#include <time.h>
#include <curses.h>
#include <signal.h>
#include <string.h>
#undef clear
#undef erase
#undef refresh

#define STANDARD_AVERAGE_WINDOW 300
#define STANDARD_DATA_FORMAT Statistics::humanReadableByte
#define STANDARD_HIDE_GRAPHS false
// #define STANDARD_MAX_DEFLECTION 10240
#define STANDARD_REFRESH_INTERVAL 500
#define STANDARD_TRAFFIC_FORMAT Statistics::humanReadableBit

using namespace std;

static OptWindow m_optWindow;
static TrafficWindow m_mainWindow;

static bool quit = false;

int main(int argc, char *argv[])
{
    SettingStore::add(Setting("AverageWindow", "Window length for average (s)", STANDARD_AVERAGE_WINDOW));
    // SettingStore::add(Setting("BarMaxIn", "Max Incoming deflection (kBit/s)", STANDARD_MAX_DEFLECTION));
    // SettingStore::add(Setting("BarMaxOut", "Max Outgoing deflection (kBit/s)", STANDARD_MAX_DEFLECTION));
    SettingStore::add(Setting("DataFormat", "Unit for data numbers", STANDARD_DATA_FORMAT));
    SettingStore::add(Setting("Devices", "Devices to show", "all"));
    SettingStore::add(Setting("MultipleDevices", "Show multiple devices", STANDARD_HIDE_GRAPHS));
    SettingStore::add(Setting("RefreshInterval", "Refresh interval (ms)", STANDARD_REFRESH_INTERVAL));
    SettingStore::add(Setting("TrafficFormat", "Unit for traffic numbers", STANDARD_TRAFFIC_FORMAT));

    SettingStore::get("AverageWindow").pushFilter(new SettingFilterMin(5));
    // SettingStore::get("BarMaxIn").pushFilter(new SettingFilterMin(10));
    // SettingStore::get("BarMaxOut").pushFilter(new SettingFilterMin(10));
    SettingStore::get("RefreshInterval").pushFilter(new SettingFilterMin(50));

    SettingStore::get("Devices").pushFilter(new SettingFilterDefault("all"));
    SettingStore::get("Devices").pushFilter(new SettingFilterExclusive("all"));

    map<string, string> valueMapping;

    valueMapping[toString(false)] = "[ ]";
    valueMapping[toString(true)] = "[x]";
    SettingStore::get("MultipleDevices").pushFilter(new SettingFilterMap(valueMapping));
    valueMapping.clear();

    valueMapping[toString(Statistics::humanReadableBit)] = "Human Readable (Bit)";
    valueMapping[toString(Statistics::humanReadableByte)] = "Human Readable (Byte)";
    valueMapping[toString(Statistics::bit)] = "Bit";
    valueMapping[toString(Statistics::byte)] = "Byte";
    valueMapping[toString(Statistics::kiloBit)] = "kBit";
    valueMapping[toString(Statistics::kiloByte)] = "kByte";
    valueMapping[toString(Statistics::megaBit)] = "MBit";
    valueMapping[toString(Statistics::megaByte)] = "MByte";
    valueMapping[toString(Statistics::gigaBit)] = "GBit";
    valueMapping[toString(Statistics::gigaByte)] = "GByte";
    SettingStore::get("TrafficFormat").pushFilter(new SettingFilterMap(valueMapping));
    SettingStore::get("DataFormat").pushFilter(new SettingFilterMap(valueMapping));
    valueMapping.clear();

    // retrieve home directory
    const char* homeDirArray = getenv("HOME");
    if(!homeDirArray)
    {
        cerr << "Could not retrieve home directory." << endl;
        exit(1);
    }

    string homeDir = homeDirArray;
    SettingStore::readFromFile(SYSCONFDIR "/nload.conf");
    SettingStore::readFromFile(homeDir + "/.nload");

    // parse the command line
    bool deleteDevicesRequested = true;
    for(int i = 1; i < argc; i++)
    {
        // does the user want help?
        if(strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0)
        {
            printHelp(false);
            exit(0);
        }
        // has the user set a non-default average time window?
        else if(strcmp(argv[i], "-a") == 0)
        {
            Setting& setting = SettingStore::get("AverageWindow");
            
            if(i < argc - 1 && isdigit(argv[ i + 1 ][0]) != 0)
            {
                setting = atoi(argv[ i + 1 ]);
                if(setting < 1)
                    setting = STANDARD_AVERAGE_WINDOW;

                i++;
            }
            else
            {
                cerr << "Wrong argument for the -a parameter." << endl;
                printHelp(true);
                exit(1);
            }
        }
        // has the user set a non-default 100% mark for
        // the incoming bandwidth bar?
        else if(strcmp(argv[i], "-i") == 0)
        {
			/*
            Setting& setting = SettingStore::get("BarMaxIn");
            
            if(i < argc - 1 && isdigit(argv[ i + 1 ][0]) != 0)
            {
                setting = atol(argv[ i + 1 ]);
                if(setting == 0)
                    setting = STANDARD_MAX_DEFLECTION;

                i++;
            }
            else*/
            {
                cerr << "Wrong argument for the -i parameter." << endl;
                printHelp(true);
                exit(1);
            }
        }
        // has the user set a non-default 100% mark for
        // the outgoing bandwidth bar?
        else if(strcmp(argv[i], "-o") == 0)
        {
			/*
            Setting& setting = SettingStore::get("BarMaxOut");
            
            if(i < argc - 1 && isdigit(argv[ i + 1 ][0]) != 0)
            {
                setting = atol(argv[ i + 1 ]);
                if(setting == 0)
                    setting = STANDARD_MAX_DEFLECTION;

                i++;
            }
            else*/
            {
                cerr << "Wrong argument for the -o parameter." << endl;
                printHelp(true);
                exit(1);
            }
        }
        // has the user set a non-default refresh interval?
        else if(strcmp(argv[i], "-t") == 0)
        {
            Setting& setting = SettingStore::get("RefreshInterval");
            
            if(i < argc - 1 && isdigit(argv[ i + 1 ][0]) != 0)
            {
                setting = atoi(argv[ i + 1 ]);
                if(setting == 0)
                    setting = STANDARD_REFRESH_INTERVAL;

                i++;
            }
            else
            {
                cerr << "Wrong argument for the -t parameter." << endl;
                printHelp(true);
                exit(1);
            }
        }
        // has the user set a non-default unit for traffic numbers?
        else if(strcmp(argv[i], "-u") == 0)
        {
            Setting& setting = SettingStore::get("TrafficFormat");
            
            if(i < argc - 1 && isalpha(argv[ i + 1 ][0]) != 0)
            {
                switch(argv[ i + 1 ][0])
                {
                    case 'H':
                        setting = Statistics::humanReadableByte;
                        break;
                    case 'h':
                        setting = Statistics::humanReadableBit;
                        break;
                    case 'B':
                        setting = Statistics::byte;
                        break;
                    case 'b':
                        setting = Statistics::bit;
                        break;
                    case 'K':
                        setting = Statistics::kiloByte;
                        break;
                    case 'k':
                        setting = Statistics::kiloBit;
                        break;
                    case 'M':
                        setting = Statistics::megaByte;
                        break;
                    case 'm':
                        setting = Statistics::megaBit;
                        break;
                    case 'G':
                        setting = Statistics::gigaByte;
                        break;
                    case 'g':
                        setting = Statistics::gigaBit;
                        break;
                    default:
                        cerr << "Wrong argument for the -u parameter." << endl;
                        printHelp(true);
                        exit(1);
                }

                i++;
            }
            else
            {
                cerr << "Wrong argument for the -u parameter." << endl;
                printHelp(true);
                exit(1);
            }
        }
        // has the user set a non-default unit for numbers of amount of data?
        else if(strcmp(argv[i], "-U") == 0)
        {
            Setting& setting = SettingStore::get("DataFormat");
            
            if(i < argc - 1 && isalpha(argv[ i + 1 ][0]) != 0)
            {
                switch(argv[ i + 1 ][0])
                {
                    case 'H':
                        setting = Statistics::humanReadableByte;
                        break;
                    case 'h':
                        setting = Statistics::humanReadableBit;
                        break;
                    case 'B':
                        setting = Statistics::byte;
                        break;
                    case 'b':
                        setting = Statistics::bit;
                        break;
                    case 'K':
                        setting = Statistics::kiloByte;
                        break;
                    case 'k':
                        setting = Statistics::kiloBit;
                        break;
                    case 'M':
                        setting = Statistics::megaByte;
                        break;
                    case 'm':
                        setting = Statistics::megaBit;
                        break;
                    case 'G':
                        setting = Statistics::gigaByte;
                        break;
                    case 'g':
                        setting = Statistics::gigaBit;
                        break;
                    default:
                        cerr << "Wrong argument for the -U parameter." << endl;
                        printHelp(true);
                        exit(1);
                }

                i++;
            }
            else
            {
                cerr << "Wrong argument for the -U parameter." << endl;
                printHelp(true);
                exit(1);
            }
        
        }
        // has the user chosen to display multiple devices and thus not to display graphs?
        else if(strcmp(argv[i], "-m") == 0)
        {
            SettingStore::get("MultipleDevices") = true;
        }
        // obsolete -b option
        else if(strcmp(argv[i], "-b") == 0)
        {
        }
        // obsolete -s option
        else if(strcmp(argv[i], "-s") == 0)
        {
        }
        // assume unknown parameter to be the network device
        else
        {
            Setting& devices = SettingStore::get("Devices");

            if(deleteDevicesRequested)
            {
                devices = "";
                deleteDevicesRequested = false;
            }

            devices.setThroughFilter(trim(devices.getThroughFilter() + " " + argv[i]));
        }
    }

    // auto-detect network devices
    DevReaderFactory::findAllDevices();
    const map<string, DevReader*>& deviceReaders = DevReaderFactory::getAllDevReaders();

    // create one instance of the Device class per device
    map<string, Device*> deviceHandlers;
    for(map<string, DevReader*>::const_iterator itDevice = deviceReaders.begin(); itDevice != deviceReaders.end(); ++itDevice)
    {
        Device* device = new Device(*itDevice->second);
        device->update();

        deviceHandlers[itDevice->first] = device;
    }

    init();

    while(!quit)
    {
        // wait RefreshInterval milliseconds (in steps of 100 ms)
        struct timespec wantedTime;
        wantedTime.tv_sec = 0;
        
        int restOfRefreshInterval = SettingStore::get("RefreshInterval");
        
        while(restOfRefreshInterval > 0 && !quit)
        {
            restOfRefreshInterval -= 100;
            wantedTime.tv_nsec = (restOfRefreshInterval >= 0 ? 100 : 100 + restOfRefreshInterval) * 1000000L;
            
            nanosleep(&wantedTime, 0);
            
            // process keyboard
            int key;
            while((key = getch()) != ERR)
            {
                if(m_optWindow.isVisible())
                {
                    if(key == KEY_F(2))
                    {
                        m_optWindow.hide();
                        m_mainWindow.resize(0, 0, Screen::width(), Screen::height());
                        restOfRefreshInterval = 0; // update the screen
                    }
                    else
                    {
                        m_optWindow.processKey(key);
                    }
                }
                else
                {
                    switch(key)
                    {
                        case KEY_F(2):
                            m_mainWindow.resize(0, Screen::height() / 4, Screen::width(), Screen::height() - Screen::height() / 4);
                            m_optWindow.show(0, 0, Screen::width(), Screen::height() / 4);
                            restOfRefreshInterval = 0; // update the screen
                            break;
                        case KEY_F(5):
                            SettingStore::writeToFile(homeDir + "/.nload");
                            break;
                        case KEY_F(6):
                            SettingStore::readFromFile("/etc/nload.conf");
                            SettingStore::readFromFile(homeDir + "/.nload");
                            break;
                        case 'q':
                        case 'Q':
                            quit = true;
                            break;
                        default:
                            m_mainWindow.processKey(key);
                    }
                }
            }
        }

        if(quit)
            break;

        vector<string> devicesRequested = split(trim(SettingStore::get("Devices").getThroughFilter()), " ");
        vector<Device*> devicesToShow;

        if(!devicesRequested.empty() && devicesRequested.front() != "all")
        {
            // check if requested devices are available
            for(vector<string>::const_iterator itRequested = devicesRequested.begin(); itRequested != devicesRequested.end(); ++itRequested)
            {
                map<string, Device*>::const_iterator itDetectedDevice = deviceHandlers.find(*itRequested);
                if(itDetectedDevice != deviceHandlers.end())
                    devicesToShow.push_back(itDetectedDevice->second);
            }
        }

        if(devicesToShow.empty())
        {
            // use all detected devices
            for(map<string, Device*>::const_iterator itDevice = deviceHandlers.begin(); itDevice != deviceHandlers.end(); ++itDevice)
                devicesToShow.push_back(itDevice->second);
        }

        // enumerate devices for display
        unsigned int deviceIndex = 0;
        for(vector<Device*>::const_iterator itDevice = devicesToShow.begin(); itDevice != devicesToShow.end(); ++itDevice)
        {
            (*itDevice)->setDeviceNumber(deviceIndex++);
            (*itDevice)->setTotalNumberOfDevices(devicesToShow.size());
        }

        // update all devices
        for(map<string, Device*>::const_iterator itDevice = deviceHandlers.begin(); itDevice != deviceHandlers.end(); ++itDevice)
            itDevice->second->update();

        // clear the screen
        m_mainWindow.clear();
        
        // print device data
        m_mainWindow.printTraffic(devicesToShow);
        
        // refresh the screen
        m_mainWindow.refresh();
        
        if(m_optWindow.isVisible())
            m_optWindow.refresh(); // always show cursor in option dialog
        
    }

    finish();
    
    for(map<string, Device*>::const_iterator itDevice = deviceHandlers.begin(); itDevice != deviceHandlers.end(); ++itDevice)
        delete itDevice->second;
    deviceHandlers.clear();
    
    return 0;
}

void init()
{
    // handle interrrupt signal
    signal(SIGINT, end);
    signal(SIGTERM, end);
    signal(SIGWINCH, terminalResized);
    
    // initialize ncurses
    initscr();
    keypad(stdscr, true);
    nodelay(stdscr, true);
    noecho();
    nonl();
    cbreak();
    
    // create main window
    m_mainWindow.show(0, 0, Screen::width(), Screen::height());
}

void finish()
{
    // destroy main window
    m_mainWindow.hide();
    
    // stop ncurses
    endwin();
}

void end(int signal)
{
    quit = true;
}

void terminalResized(int signal)
{
    bool optWindowWasVisible = m_optWindow.isVisible();

    m_optWindow.hide();

    finish();   
    init();
    
    if(optWindowWasVisible)
    {
        m_mainWindow.resize(0, Screen::height() / 4, Screen::width(), Screen::height() - Screen::height() / 4);
        m_optWindow.show(0, 0, Screen::width(), Screen::height() / 4);
    }
}

void printHelp(bool error)
{
    // print disclaimer
    (error ? cerr : cout)
        << "\n"
        << PACKAGE << " version " << VERSION << "\n"
        << "Copyright (C) 2001 - 2012 by Roland Riegel <feedback@roland-riegel.de>\n"
        << PACKAGE << " comes with ABSOLUTELY NO WARRANTY. This is free software, and you are\n"
        << "welcome to redistribute it under certain conditions. For more details see the\n"
        << "GNU General Public License Version 2 (http://www.gnu.org/copyleft/gpl.html).\n\n"

        << "Command line syntax:\n"
        << PACKAGE << " [options] [devices]\n"
        << PACKAGE << " --help|-h\n\n"

        << "Options:\n"
        << "-a period       Sets the length in seconds of the time window for average\n"
        << "                calculation.\n"
        << "                Default is " << STANDARD_AVERAGE_WINDOW << ".\n"
        // << "-i max_scaling  Specifies the 100% mark in kBit/s of the graph indicating the\n"
        // << "                incoming bandwidth usage. Ignored if max_scaling is 0 or the\n"
        // << "                switch -m is given.\n"
        // << "                Default is " << STANDARD_MAX_DEFLECTION << ".\n"
        << "-m              Show multiple devices at a time; no traffic graphs.\n"
        // << "-o max_scaling  Same as -i but for the graph indicating the outgoing bandwidth\n"
        // << "                usage.\n"
        // << "                Default is " << STANDARD_MAX_DEFLECTION << ".\n"
        << "-t interval     Determines the refresh interval of the display in milliseconds.\n"
        << "                Default is " << STANDARD_REFRESH_INTERVAL << ".\n"
        << "-u h|b|k|m|g    Sets the type of unit used for the display of traffic numbers.\n"
        << "   H|B|K|M|G    h: auto, b: Bit/s, k: kBit/s, m: MBit/s etc.\n"
        << "                H: auto, B: Byte/s, K: kByte/s, M: MByte/s etc.\n"
        << "                Default is h.\n"
        << "-U h|b|k|m|g    Same as -u, but for a total amount of data (without \"/s\").\n"
        << "   H|B|K|M|G    Default is H.\n"
        << "devices         Network devices to use.\n"
        << "                Default is to use all auto-detected devices.\n"
        << "--help\n"
        << "-h              Print this help.\n\n"

        << "example: " << PACKAGE << " -t 200 -i 1024 -o 128 -U M\n\n"

        << "The options above can also be changed at run time by pressing the 'F2' key.\n"
		<< "-i/-o			deprecated parameter.NOT use this.\n"
		<< "				now support dynamic scaling.\n"
        << endl;
}

